### DISCUSSION TITLE
Validating array with numerical keys

### DISCUSSION THREAD

**Discussion by ukenpachi** [laravel/framework#39257] [Q&A] [answered]
- Lumen Version: 8.2.3
- PHP Version: 8
- Database Driver & Version: mysql

### Description:
```
  Use Validator;

  $rules = ['ids.*' => 'required|integer'];
  $ids = json_decode($request->get('ids'), true);

  $validator = Validator::make($ids, $rules);

  if ($validator->fails()) {
      $this->response->httpCode = 400;
      $this->response->success = false;
      $this->response->addError($this->errorCode . '0001', $this->errorCode . '0001', $validator->errors()->all());
      return $this->json($this->response);
  }
```
ids = `["2226","4307","29444","1180","sfdsdfjdfdjsf"] ` validation does not failed with this example json, it is because the Validator does not seems to work out when array does not have non numeric key. Any ideas how to make this work using validator? thank you

### Steps To Reproduce:
**Please note that $ids is a json string** 
$ids example is ` "["2226","4307","??-$?","1180","sfdsdfjdfdjsf"]"`

**Comment by jmarcher** [comment]
You are doing it wrong, you should pass the request or the whole request data into the make() method.

Your rule checks for the `ids` key but you are just plucking it from the request (which wont have the `ids` key)

`  $validator = Validator::make($request, $rules);` should work

**Reply by ukenpachi** [reply]
> request

Thanks for taking interest into this discussion.  I did tried that but I got this error message

`Use  Illuminate\Http\Request`

`Illuminate\Validation\Factory::make(): Argument #1 ($data) must be of type array, Laravel\Lumen\Http\Request given, called in /home/vagrant/code/api/vendor/illuminate/support/Facades/Facade.php on line 261`

**Comment by jmarcher** [comment]
Then use `$request->all()` instead

**Reply by ukenpachi** [reply]
> Then use `$request->all()` instead

tried that already still the same thing.

**Comment by ukenpachi** [comment]
If I pass the array the validation rules does not work at all. It just automatically passes it without checking against the rule.
In the image you can see it at the new breakpoint which it shouldn't be because the $ids array has letters in one of the values, this means it should fail validation but it does not.
![image](https://user-images.githubusercontent.com/5822229/137904796-7b796eca-b83e-4e73-a1db-93a9223b43eb.png)

**Comment by jmarcher** [comment]
Validator make still points the wrong data. And required|integer does not make much sense you could do something like

```
['ids' => 'required', 'ids.*' => 'integer']
```

**Comment by dennisprudlo** [comment, marked as answer]
```php
$validator = Validator::make($request->only('ids'), [
    'ids' => 'required|array',
    'ids.*' => 'integer'
]);
```

Here are the differences between `all`, `only`, and `get`:
```php
// Let's assume we have the following request data:
// name => 'John Doe'
// admin => true
// ids => [1, 2, 3, 4, 5]

request()->all()
// [
//     'name' => 'John Doe',
//     'admin' => true,
//     'ids' => [1, 2, 3, 4, 5]
// ]

request()->only('ids')
// [
//     'ids' => [1, 2, 3, 4, 5]
// ]

request()->get('ids')
// [1, 2, 3, 4, 5]
```

When you use `get` you receive the value at the specified key. But what you really want is to get the key-value pair for the specified key.

**Reply by ukenpachi** [reply]
> ```
> $validator = Validator::make($request->only('ids'), [
>     'ids' => 'required|array',
>     'ids.*' => 'integer'
> ]);
> ```
> 
> Here are the differences between `all`, `only`, and `get`:
> 
> ```
> // Let's assume we have the following request data:
> // name => 'John Doe'
> // admin => true
> // ids => [1, 2, 3, 4, 5]
> 
> request()->all()
> // [
> //     'name' => 'John Doe',
> //     'admin' => true,
> //     'ids' => [1, 2, 3, 4, 5]
> // ]
> 
> request()->only('ids')
> // [
> //     'ids' => [1, 2, 3, 4, 5]
> // ]
> 
> request()->get('ids')
> // [1, 2, 3, 4, 5]
> ```
> 
> When you use `get` you receive the value at the specified key. But what you really want is to get the key-value pair for the specified key.

![image](https://user-images.githubusercontent.com/5822229/137921439-d6eae1ab-565d-438c-9433-c06b0f2aee14.png)

still validation get skipped. Do bare in mind that my ids are json string. In the image I added $k to show you what ->only('ids') looks like

**Reply by dennisprudlo** [reply]
Aaah I see ok, then you additionally need to `json_decode` your string like you did before.

```php
$data = [
    'ids' => json_decode($request->get('ids'))
];

validator($data, $rules)->fails(); // true
```

**Reply by ukenpachi** [reply]
> Aaah I see ok, then you additionally need to `json_decode` your string like you did before.
> 
> ```
> $data = [
>     'ids' => json_decode($request->get('ids'))
> ];
> 
> validator($data, $rules)->fails(); // true
> ```

thank you. this worked, can you put it, on it own and I will mark this as the answer. Thanks again

**Reply by dennisprudlo** [reply]
No need. If you want to "close" this question just mark the first comment.

**Comment by jmarcher** [comment]
With `request()->get('ids')` you get the  key-value pair, PHP will not print the keys because they are correlational

When test the following:

`[1, "a", "b", 1]` wont show any keys but if you insert one key which does not follow the order then all the keys will be shown: `[1, "a", "b", 100=>1]`

**Reply by ukenpachi** [reply]
> With `request()->get('ids')` you get the key-value pair, PHP will not print the keys because they are correlational
> 
> When test the following:
> 
> `[1, "a", "b", 1]` wont show any keys but if you insert one key which does not follow the order then all the keys will be shown: `[1, "a", "b", 100=>1]`

looks like it not take same has my ids is a json string
![image](https://user-images.githubusercontent.com/5822229/137922598-e3818d52-af58-4782-adf4-885a85a1e74e.png)

**Reply by jmarcher** [reply]
You did nothing of what we suggested, the code you show is still wrong.
